rclcpp/test/rclcpp/test_generic_client.cpp

```diff
@@ -0,0 +1,463 @@
+  auto server_node = std::make_shared<rclcpp::Node>("test_server_node", "ns");
+
+  rclcpp::ServicesQoS client_qos;
+  client_qos.keep_last(3);
+  auto client =
+    client_node->create_generic_client("test_service", "test_msgs/srv/Empty", client_qos);
+  std::atomic<size_t> server_requests_count {0};
+  auto server_callback = [&server_requests_count](
+    const test_msgs::srv::Empty::Request::SharedPtr,
+    test_msgs::srv::Empty::Response::SharedPtr) {server_requests_count++;};
+  auto server = server_node->create_service<test_msgs::srv::Empty>(
+    "test_service", server_callback, client_qos);
+  auto request = std::make_shared<test_msgs::srv::Empty::Request>();
+
+  std::atomic<size_t> c1 {0};
+  auto increase_c1_cb = [&c1](size_t count_msgs) {c1 += count_msgs;};
+  client->set_on_new_response_callback(increase_c1_cb);
+
+  client->async_send_request(request.get());
+  auto start = std::chrono::steady_clock::now();
+  while (server_requests_count == 0 &&
+    (std::chrono::steady_clock::now() - start) < 10s)
+  {
+    rclcpp::spin_some(server_node);
+  }
+
+  ASSERT_EQ(server_requests_count, 1u);
+
+  start = std::chrono::steady_clock::now();
+  do {
+    std::this_thread::sleep_for(100ms);
+  } while (c1 == 0 && std::chrono::steady_clock::now() - start < 10s);
+
+  EXPECT_EQ(c1.load(), 1u);
+
+  std::atomic<size_t> c2 {0};
+  auto increase_c2_cb = [&c2](size_t count_msgs) {c2 += count_msgs;};
+  client->set_on_new_response_callback(increase_c2_cb);
+
+  client->async_send_request(request.get());
+  start = std::chrono::steady_clock::now();
+  while (server_requests_count == 1 &&
+    (std::chrono::steady_clock::now() - start) < 10s)
+  {
+    rclcpp::spin_some(server_node);
+  }
+
+  ASSERT_EQ(server_requests_count, 2u);
+
+  start = std::chrono::steady_clock::now();
+  do {
+    std::this_thread::sleep_for(100ms);
+  } while (c1 == 0 && std::chrono::steady_clock::now() - start < 10s);
+
+  EXPECT_EQ(c1.load(), 1u);
+  EXPECT_EQ(c2.load(), 1u);
+
+  client->clear_on_new_response_callback();
+
+  client->async_send_request(request.get());
+  client->async_send_request(request.get());
+  client->async_send_request(request.get());
+  start = std::chrono::steady_clock::now();
+  while (server_requests_count < 5 &&
+    (std::chrono::steady_clock::now() - start) < 10s)
+  {
+    rclcpp::spin_some(server_node);
+  }
+
+  ASSERT_EQ(server_requests_count, 5u);
+
+  std::atomic<size_t> c3 {0};
+  auto increase_c3_cb = [&c3](size_t count_msgs) {c3 += count_msgs;};
+  client->set_on_new_response_callback(increase_c3_cb);
+
+  start = std::chrono::steady_clock::now();
+  do {
+    std::this_thread::sleep_for(100ms);
+  } while (c3 < 3 && std::chrono::steady_clock::now() - start < 10s);
+
+  EXPECT_EQ(c1.load(), 1u);
+  EXPECT_EQ(c2.load(), 1u);
+  EXPECT_EQ(c3.load(), 3u);
+
+  std::function<void(size_t)> invalid_cb = nullptr;
+  EXPECT_THROW(client->set_on_new_response_callback(invalid_cb), std::invalid_argument);
+}
```

Comment:
Similar comment here, let's not copy-paste tests (this is from https://github.com/ros2/rclcpp/blob/126d517193b6df8177680898e23906a85792eaf6/rclcpp/test/rclcpp/test_client.cpp#L413-L503 and it's not just this test, it's most of this file) that are hugely complex, instead please refactor them to have the same code run over Generic and regular Clients.